lightning/src/util/scid_utils.rs

```diff
@@ -60,6 +60,176 @@ pub fn scid_from_parts(block: u64, tx_index: u64, vout_index: u64) -> Result<u64
+			where K::Target: KeysInterface<Signer = Signer>,
+		{
+			// Ensure we haven't created a namespace that doesn't fit into the 3 bits we've allocated for
+			// namespaces.
+			assert!((*self as u8) < MAX_NAMESPACES);
+			const BLOCKS_PER_MONTH: u32 = 144 /* blocks per day */ * 30 /* days per month */;
+			let rand_bytes = keys_manager.get_secure_random_bytes();
+
+			let segwit_activation_height = segwit_activation_height(genesis_hash);
+			let mut valid_block_range = if highest_seen_blockheight > segwit_activation_height {
+				highest_seen_blockheight - segwit_activation_height
+			} else {
+				1
+			};
+			// We want to ensure that this fake channel won't conflict with any transactions we haven't
+			// seen yet, in case `highest_seen_blockheight` is updated before we get full information
+			// about transactions confirmed in the given block.
+			if valid_block_range > BLOCKS_PER_MONTH { valid_block_range -= BLOCKS_PER_MONTH; }
+
+			let rand_for_height = u32::from_be_bytes(rand_bytes[..4].try_into().unwrap());
+			let fake_scid_height = segwit_activation_height + rand_for_height % valid_block_range;
+
+			let rand_for_tx_index = u32::from_be_bytes(rand_bytes[4..8].try_into().unwrap());
+			let fake_scid_tx_index = rand_for_tx_index % MAX_TX_INDEX;
+
+			// Put the scid in the given namespace.
+			let fake_scid_vout = self.get_encrypted_vout(fake_scid_height, fake_scid_tx_index, fake_scid_rand_bytes);
+			scid_utils::scid_from_parts(fake_scid_height as u64, fake_scid_tx_index as u64, fake_scid_vout as u64).unwrap()
+		}
+
+		/// We want to ensure that a 3rd party can't identify a payment as belong to a given
+		/// `Namespace`. Therefore, we encrypt it using a random bytes provided by `ChannelManager`.
+		fn get_encrypted_vout(&self, block_height: u32, tx_index: u32, fake_scid_rand_bytes: &[u8; 32]) -> u8 {
+			let mut salt = [0 as u8; 8];
+			let block_height_bytes = block_height.to_be_bytes();
+			salt[0..4].copy_from_slice(&block_height_bytes);
+			let tx_index_bytes = tx_index.to_be_bytes();
+			salt[4..8].copy_from_slice(&tx_index_bytes);
+
+			let mut chacha = ChaCha20::new(fake_scid_rand_bytes, &salt);
+			let mut vout_byte = [*self as u8];
+			chacha.process_in_place(&mut vout_byte);
+			vout_byte[0] & NAMESPACE_ID_BITMASK
+		}
+	}
+
+	pub fn get_phantom_scid<Signer: Sign, K: Deref>(fake_scid_rand_bytes: &[u8; 32], highest_seen_blockheight: u32, genesis_hash: &BlockHash, keys_manager: &K) -> u64
+		where K::Target: KeysInterface<Signer = Signer>,
+	{
+		let namespace = Namespace::Phantom;
+		namespace.get_fake_scid(highest_seen_blockheight, genesis_hash, fake_scid_rand_bytes, keys_manager)
+	}
+
+	fn segwit_activation_height(genesis: &BlockHash) -> u32 {
+		const MAINNET_GENESIS_STR: &'static str = "000000000019d6689c085ae165831e934ff763ae46a2a6c172b3f1b60a8ce26f";
+		if BlockHash::from_hex(MAINNET_GENESIS_STR).unwrap() == *genesis {
+			MAINNET_SEGWIT_ACTIVATION_HEIGHT
```

Comment:
Do we care if this isn't accurate for other `Network`s? Guessing no.

Reply:
Worst case it means (very marginally) less privacy, so dunno why we would, really.